src/platform/common/providerBasedQuickPick.ts

```diff
@@ -0,0 +1,407 @@
+    onDidChange: Event<void>;
+    onDidChangeStatus: Event<void>;
+    readonly items: T[];
+    readonly status: 'discovering' | 'idle';
+    refresh: () => Promise<void>;
+}
+interface SeparatorQuickPickItem extends QuickPickItem {
+    isEmptyCondaEnvironment?: boolean;
+}
+class CommandQuickPickItem<T extends { id: string }> extends BaseQuickPickItem {
+    constructor(
+        label: string,
+        public readonly execute: () => Promise<T | undefined | typeof InputFlowAction.back>
+    ) {
+        super(label);
+    }
+}
+function isSelectorQuickPickItem<T extends { id: string }>(item: QuickPickItem): item is SelectorQuickPickItem<T> {
+    return item instanceof SelectorQuickPickItem;
+}
+
+export class BaseProviderBasedQuickPick<T extends { id: string }> extends Disposables {
+    private readonly categories = new Map<QuickPickItem, Set<SelectorQuickPickItem<T>>>();
+    private quickPickItems: QuickPickItem[] = [];
+    private readonly quickPick: QuickPick<QuickPickItem>;
+    private readonly provider: IQuickPickItemProvider<T>;
+    private readonly token: CancellationToken;
+
+    constructor(
+        private readonly options: {
+            provider: IQuickPickItemProvider<T>;
+            token: CancellationToken;
+            placeholder?: string;
+            supportsBack: boolean;
+            isSelected?: (item: T) => boolean;
+            isRecommended?: (item: T) => boolean;
+            toQuickPick: (item: T) => SelectorQuickPickItem<T>;
+            getCategory: (item: T) => { label: string; sortKey?: string };
+        }
+    ) {
+        super();
+        this.provider = options.provider;
+        this.token = options.token;
+        const refreshButton: QuickInputButton = { iconPath: new ThemeIcon('refresh'), tooltip: Common.refresh };
+        const refreshingButton: QuickInputButton = {
+            iconPath: new ThemeIcon('loading~spin'),
+            tooltip: Common.refreshing
+        };
+        const quickPick = (this.quickPick = window.createQuickPick());
+        this.disposables.push(quickPick);
+        quickPick.title = this.provider.title;
+        quickPick.placeholder = options.placeholder || '';
+        quickPick.items = this.getAdditionalQuickPickItems().concat(this.quickPickItems);
+        quickPick.buttons = options.supportsBack ? [QuickInputButtons.Back, refreshButton] : [refreshButton];
+        quickPick.ignoreFocusOut = true;
+        quickPick.busy = this.provider.status === 'discovering';
+        quickPick.onDidTriggerButton(
+            async (e) => {
+                if (e === refreshButton) {
+                    const oldButtons = quickPick.buttons;
+                    quickPick.buttons = quickPick.buttons
+                        .filter((btn) => btn !== refreshButton)
+                        .concat(refreshingButton);
+                    await this.provider.refresh().catch(noop);
+                    quickPick.buttons = oldButtons;
+                }
```

Comment:
Is this replacing the `loading` icon button with the `loading~spin` icon? If so, is this a common pattern that we want to use? I think just `busy = true` is what we suggest all extension use. @TylerLeonhardt in case you want to chime in as the quick pick owner.

Reply:
Pattern aside, it might make more sense to use `map` to preserve button order: 

```suggestion
                if (e === refreshButton) {
                    const oldButtons = quickPick.buttons;
                    quickPick.buttons = quickPick.buttons
                        .map((btn) => btn !== refreshButton ? btn : refreshingButton);
                    await this.provider.refresh().catch(noop);
                    quickPick.buttons = oldButtons;
                }
```

Reply:
Agreed makes sense, will use `busy=true`
Not sure why we did this in the past (this is code that I copied)
Will merge this PR and makde that change in a smaller PR, thanks